### PR TITLE
[PIE-932] preview of the iterator interface to dict reader 

### PR DESCRIPTION
**Pull request by guyd**
Added support to read file from CSV. Also, added the support for canonical words

### CODE REVIEW

File: spellchecker/symspell.py [hunk 1]
```diff
@@ -74,6 +74,8 @@ def create_dictionary_entry(self, key, count):
         Keyword arguments:
         key -- The word to add to dictionary.
         count -- The frequency count for word.
+        canonical_term -- This is an optional term (or multi-terms) that will used as suggestion instead of
```
Comment (ajutras) [inline spellchecker/symspell.py]: if you add to the docstring can you format it do google style ?

File: spellchecker/symspell.py [hunk 2]
```diff
@@ -62,7 +62,7 @@ def __init__(self, initial_capacity=16, max_dictionary_edit_distance=2,
         self._distance_algorithm = DistanceAlgorithm.DAMERUAUOSA
         self._max_length = 0
 
-    def create_dictionary_entry(self, key, count):
+    def create_dictionary_entry(self, key, count, canonical_term=None):
```
Comment (ajutras) [inline spellchecker/symspell.py]: don't forget typehint please (everywhere in your new code)

File: spellchecker/symspell.py [hunk 3]
```diff
@@ -105,20 +107,38 @@ def create_dictionary_entry(self, key, count):
...
             count = (count_previous + count
                      if sys.maxsize - count_previous > count
                      else sys.maxsize)
-            self._words[key] = count
+
+            def get_canonical_term(key, former_canonical_term, canonical_term):
+                """
+                If there are many conflicting canonical terms for the same key, use the latest non-null one or
+                the key value.
+                Args:
```
Comment (ajutras) [inline spellchecker/symspell.py]: blank line before Args

File: spellchecker/symspell.py [hunk 4]
```diff
@@ -137,28 +157,21 @@ def create_dictionary_entry(self, key, count):
...
+        """
+        Load multiple dictionary entries from a file of word/frequency
         count pairs. Merges with any dictionary data already loaded.
 
-        Keyword arguments:
-        corpus -- The path+filename of the file.
-        term_index -- The column position of the word.
-        count_index -- The column position of the frequency count.
+        Args:
+            word_iterator:
```
Comment (ajutras) [inline spellchecker/symspell.py]: just a reminder to complete docstring

File: spellchecker/symspell.py (hunk 2 above)
Comment (guyd) [reply spellchecker/symspell.py]: Completely forgot about those. Thanks for reminding me.

File: spellchecker/symspell.py [hunk 5]
```diff
@@ -137,29 +161,20 @@ def create_dictionary_entry(self, key, count):
                 self._deletes[delete_hash] = [key]
         return True
 
-    def load_dictionary(self, corpus, term_index, count_index):
-        """Load multiple dictionary entries from a file of word/frequency
+    def load_dictionary(self, word_iterator: Iterable[tuple]) -> None:
```
Comment (ajutras) [inline spellchecker/symspell.py]: Whenever a function returns nothing you can omit the typehint return value as well as the Returns part of the docstring

File: spellchecker/helpers.py [hunk 6]
```diff
@@ -70,3 +72,84 @@ def is_acronym(word):
...
+                count = try_parse_int64(line_parts[self.count_index])
+                if count is not None:
+                    canonical_term = line_parts[self.canonical_term_index] if self.canonical_term_index else None
+                    return term, count, canonical_term
+        raise StopIteration
+
+
+class CsvFileIterator:
+    """
+       Iterate over a CSV file.
```
Comment (ajutras) [inline spellchecker/helpers.py]: docstring indented weirdly, don't forget to typehint the __init__ for both classes

File: spellchecker/symspell.py [hunk 7]
```diff
@@ -62,24 +63,28 @@ def __init__(self, initial_capacity=16, max_dictionary_edit_distance=2,
...
         create_dictionary_entry
 
-        Keyword arguments:
-        key -- The word to add to dictionary.
-        count -- The frequency count for word.
+       Args:
+           key: The word to add to dictionary.
+           count: The frequency count for word.
+           canonical_term: This is an optional term (or multi-terms) that will used as suggestion instead of
+        the original word (e.g. Saint-Laurent instead of St-Laurent). By default, the key value is used
```
Comment (ajutras) [inline spellchecker/symspell.py]: Soft comments:

if the description of a param requires more than one line indent the following lines (see: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

Also the text of your Returns is not indented by 4.

**Review by ajutras — APPROVED** [review]

**Pull request merged**